keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,480 @@
+type Root struct {
+  Entry SubLevel
+}
+
+type SubLevel struct {
+  // +defaut=default-name
+  Name string
+  // +default=0
+  Number int
+}
+```
+
+We would generate the following OpenAPI definition, both for CRD types (by kubebuilder)
+and built-in types:
+```yaml
+Root:
+  type: object
+  default: {}
+  properties:
+    entry:
+      $ref: SubLevel
+      default: {}
+SubLevel:
+  type: object
+  properties:
+    name:
+      type: string
+      default: default-name
+    number:
+      type: integer
+      default: 0
+```
+
+which would have the following behavior, both for CRD types and built-in types:
+```python
+>>> default('null')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: null}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {}}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+```
+
+#### Struct pointers
+
+Struct pointers can be have a specific default since they are distinct
+from unspecified non-pointers structs. Modifying the type from the
+previous example:
+```golang
+type Root struct {
+  // This is now a pointer, we can provide a default:
+  // +default={name: pointer-name}
+  Entry *SubLevel
+}
+
+type SubLevel struct {
+  // +defaut=default-name
+  Name string
+  // +default=0
+  Number int
+}
+```
+
+We would generate the following OpenAPI definition, both for CRD types (by kubebuilder)
+and built-in types:
+```yaml
+Root:
+  type: object
+  default: {}
+  properties:
+    entry:
+      $ref: SubLevel
+      default: {name: pointer-name}
+SubLevel:
+  type: object
+  properties:
+    name:
+      type: string
+      default: default-name
+    number:
+      type: integer
+      default: 0
+```
+
+which would have the following behavior, both for CRD types and built-in types:
+```python
+>>> default('null')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{}')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{entry: null}')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{entry: {}}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+```
+
+### Test Plan
+
+As soon as this is implemented and working, we can start removing a lot
+of the existing defaulting code and replace it with this values, and we
+can re-use the existing defaulting tests, make sure that the behavior
+doesn't change when moving from code-based defaults to declarative
+default. Also panic in specific places of the defaulters to guarantee
+that the defaulting have happened before.
+
+### Graduation Criteria
+
+The feature will be released as Beta and enabled by default. This is to
+give a chance to disable the feature if something goes wrong, but we're
+expecting GA quality from the start, with no changes.
```

Comment:
why not alpha and disabled by default?

Reply:
I think it's very confusing if we have `// +default=...` in the built-in types comments but it's not honored because the feature is alpha (and depends on an apiserver flag).